### PR TITLE
build: migrate to eslint flat config format

### PR DESCRIPTION
**Pull request by ollien**
Modern versions of eslint have migrated to the ["flat" configuration style](https://eslint.org/docs/latest/use/configure/migration-guide), deprecating `.eslintrc.js`. I've gone ahead and migrated the project to the new format, which will allow us to safely upgrade to newer versions of eslint. I believe I've preserved all existing rules, but it's a bit hard to prove a negative :sweat_smile: 

The `globals` dependency is needed due to a change in how the new configuration format handles known globals. See: https://eslint.org/docs/latest/use/configure/migration-guide#configuring-language-options

### CODE REVIEW

**Comment by xiyaowong** [comment]
Nice. I've been wanting to do this for a long time. 

Also need to turn on "eslint.experimental.useFlatConfig"

https://github.com/vscode-neovim/vscode-neovim/blob/1fc6752429e1326f98557bd7a5961030e935b5f7/.vscode/settings.json#L14

File: eslint.config.cjs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            "@typescript-eslint/explicit-function-return-type": "off",
+            "@typescript-eslint/no-non-null-assertion": "off",
+            "@typescript-eslint/no-explicit-any": "off",
+        },
+    },
+    {
+        files: ["src/test/**/*.test.ts"],
+        languageOptions: {
+            globals: {
+                ...globals.jest,
```
Comment (xiyaowong) [inline eslint.config.cjs:74]: We are not using jest

Comment (ollien) [reply eslint.config.cjs:74]: Good callout. I took ths from the old config without realizing this. Looks like there's some `mocha` globals we can use.

Comment (xiyaowong) [reply eslint.config.cjs]: :joy: Okay, I haven’t looked at eslintrc carefully.

**Comment by xiyaowong** [comment]
I don't understand why it's not "chore" but "build".

**Comment by ollien** [comment]
I can change the title if you want. I know there were discussions of not using `chore` so I picked the next best thing. I have no strong opinions one way or the other

**Comment by xiyaowong** [comment]
😵

**Comment by justinmk** [comment]
"chore" is a useless prefix. Of course this is either "build" or "lint". Why wouldn't it be?

**Review by justinmk — APPROVED** [review]

**Comment by xiyaowong** [comment]
Should delete `.eslintignore`

**Comment by ollien** [comment]
> "chore" is a useless prefix. Of course this is either "build" or "lint". Why wouldn't it be?

FYI, `lint` is not a prefix we allow. From the CI:

`allowed-commit-types: feat,fix,docs,style,refactor,test,build,perf,ci,chore,revert,merge,wip`

I agree that would be the most fitting here

**Comment by justinmk** [comment]
> `lint` is not a prefix we allow.

we can add it (and remove "chore")

**Review by theol0403 — APPROVED** [review]

**Comment by xiyaowong** [comment]
> we can add it

👍🏻

 (and remove "chore")

👎🏻

Most front-end projects use something like chore(lint)

**Comment by justinmk** [comment]
> Most front-end projects use something like chore(lint)

Why does that matter? 

The problem is that people over-use "chore", it ends up being 30% of the commits.

**Pull request merged**